src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerations.cs

```diff
@@ -0,0 +1,187 @@
+        {
+            var wellKnowTypeProvider = WellKnownTypeProvider.GetOrCreate(context.Compilation);
+            var cfg = context.OperationBlocks.GetControlFlowGraph();
+            if (cfg == null)
+            {
+                return;
+            }
+
+            using var blockToTargetInvocationOperationsMapBuilder = PooledDictionary<BasicBlock, ImmutableArray<IInvocationOperation>>.GetInstance();
+            var basicBlocks = cfg.Blocks;
+            foreach (var block in basicBlocks)
+            {
+                if (!block.IsReachable || block.Operations.IsEmpty)
+                {
+                    continue;
+                }
+
+                using var arrayBuilder = ArrayBuilder<IInvocationOperation>.GetInstance();
+                foreach (var operation in block.Operations)
+                {
+                    var immediateExecutedInvocationOperations = GetAllImmediateExecutedInvocationOperations(operation, iEnumerableType);
+                    if (immediateExecutedInvocationOperations.IsEmpty)
+                    {
+                        continue;
+                    }
+
+                    arrayBuilder.AddRange(immediateExecutedInvocationOperations);
+                }
+
+                blockToTargetInvocationOperationsMapBuilder.Add(block, arrayBuilder.ToImmutable());
+            }
+
+            if (blockToTargetInvocationOperationsMapBuilder.Count == 0)
+            {
+                return;
+            }
+
+            var blockToTargetInvocationOperationsMap = blockToTargetInvocationOperationsMapBuilder.ToImmutableDictionary();
+            var result = InvocationCountAnalysis.TryGetOrComputeResult(
+                cfg,
+                context.OwningSymbol,
+                wellKnowTypeProvider,
+                context.Options,
+                MultipleEnumerableDescriptor,
+                pessimisticAnalysis: false,
+                trackingMethodNames: s_executedImmediateMethods,
+                cancellationToken: context.CancellationToken);
+            if (result == null)
+            {
+                return;
+            }
+
+            foreach (var (_, invocationOperations) in blockToTargetInvocationOperationsMap)
+            {
+                // TODO: Test IEnumerable[]
+                var multipleEnumerationTargetSet = invocationOperations.WhereAsArray(InvokedMoreThanOneTime)
+                    .Select(op => op.Arguments[0].Parameter)
+                    .ToImmutableHashSet();
+                foreach (var operation in invocationOperations)
+                {
+                    if (multipleEnumerationTargetSet.Contains(operation.Arguments[0].Parameter))
+                    {
+                        context.ReportDiagnostic(operation.Arguments[0].CreateDiagnostic(MultipleEnumerableDescriptor));
+                    }
+                }
+            }
+
+            bool InvokedMoreThanOneTime(IInvocationOperation invocationOperation)
```

Comment:
Could cache the delegate here so it's not allocated in a loop in `multipleEnumerationTargetSet = WhereAsArray`

Reply:
Shouldn't `InvokedMoreThanOneTime` be `static`?

if  `InvokedMoreThanOneTime` is `static`, the compiler will generate code to cache the delegate for this:

```csharp
var multipleEnumerationTargetSet = invocationOperations.WhereAsArray(static io => InvokedMoreThanOneTime(io))
```

or this:

```csharp
var multipleEnumerationTargetSet = invocationOperations.WhereAsArray(static invocationOperation => {
    var arguments = invocationOperation.Arguments;
    if (!arguments.IsEmpty)
    {
        var count = result[invocationOperation.Kind, invocationOperation.Syntax];
        return count.Kind == InvocationCountAbstractValueKind.MoreThanOneTime;
    }

    return false;
    }
})
```

Reply:
@paulomorgado there's no way to get result inside the static context :D

Reply:
Ohhh, okay: https://sourceroslyn.io/#Microsoft.CodeAnalysis/Collections/ImmutableArrayExtensions.cs,332

This could work:
```csharp
var multipleEnumerationTargetSet = invocationOperations.WhereAsArray(static (invocationOperation, result) => {
    var arguments = invocationOperation.Arguments;
    if (!arguments.IsEmpty)
    {
        var count = result[invocationOperation.Kind, invocationOperation.Syntax];
        return count.Kind == InvocationCountAbstractValueKind.MoreThanOneTime;
    }

    return false;
    }
}, result);
```

Reply:
The reason I was saying caching without explicitly going for the static version is that you could theoretically cache a `readonly Func<IInvocationOperation, bool>` on the instance without forcing it to be static.

Reply:
Thanks for the advice here : )
Will address it later